lib/connector/OAuthConnector.js

```diff
@@ -182,52 +227,90 @@ class OAuthConnector {
-                if (!isNaN(userContext.vendorToken.expires_in)) {
-                    userContext.vendorToken.expires_at = Date.now() + +userContext.vendorToken.expires_in * 1000;
-                }
-                userContext.vendorUserProfile = await this.getUserProfile(userContext.vendorToken);
-                userContext.status = 'authenticated';
-                await this.saveUser(fusebitContext, userContext);
+    async ensureAccessToken(fusebitContext, userContext, foreignVendorId) {
+        if (foreignVendorId) {
+            // Get access token from foreign OAuth connector specified in userContext.foreignOAuthIdentities
+            const oauthIdentity = (userContext.foreignOAuthIdentities || {})[foreignVendorId];
+            Sdk.debug('OBTAINING ACCESS TOKEN FOR FOREIGN USER', foreignVendorId, oauthIdentity);
+            if (oauthIdentity) {
+                const response = await Superagent.get(
+                    `${oauthIdentity.connectorBaseUrl}/user/${encodeURIComponent(oauthIdentity.userId)}/token`
+                ).set('Authorization', `Bearer ${fusebitContext.fusebit.functionAccessToken}`);
+                return response.body;
+            } else {
+                throw new Error(
+                    `The user ${userContext.vendorUserId} is not associated with an identity in the ${foreignVendorId} OAuth connector.`
+                );
+            }
+        } else {
+            // Get access token for "this" OAuth connector
+            if (userContext.vendorToken.access_token && userContext.vendorToken.expires_at > Date.now() + 30000) {
+                Sdk.debug('RETURNING CURRENT ACCESS TOKEN FOR USER', userContext.vendorUserId);
                 return userContext.vendorToken;
-            } catch (e) {
-                await this.deleteUser(fusebitContext, userContext.vendorUserId);
-                Sdk.debug('REFRESH TOKEN ERROR', e);
+            }
+            if (userContext.vendorToken.refresh_token) {
+                Sdk.debug('REFRESHING ACCESS TOKEN FOR USER', userContext.vendorUserId);
+                userContext.status = 'refreshing';
+                try {
+                    await this.saveUser(fusebitContext, userContext);
+                    userContext.vendorToken = await this.refreshAccessToken(fusebitContext, userContext.vendorToken);
+                    if (!isNaN(userContext.vendorToken.expires_in)) {
+                        userContext.vendorToken.expires_at = Date.now() + +userContext.vendorToken.expires_in * 1000;
+                    }
+                    userContext.vendorUserProfile = await this.getUserProfile(userContext.vendorToken);
+                    userContext.status = 'authenticated';
+                    await this.saveUser(fusebitContext, userContext);
+                    return userContext.vendorToken;
+                } catch (e) {
+                    await this.deleteUser(fusebitContext, userContext.vendorUserId);
```

Comment:
Deleting a user seems a touch harsh, given the several calls to saveUser that could also have failed here, as well as the other possibly-still-valid foreign vendor oauth's that might be valid.

Reply:
Enhanced the refresh logic with: 
1. Addressing the herd problem. First client initiates the refresh operation, the rest of them waits until it completes with exponential backoff and limit. 
2. Only deleting the client if a configurable limit of consecutive refresh failures was reached (by default 10).

Reply:
I’m looking at this on my iPad - can you point me to the herd back off logic?  I’ve only managed to find on-failure back off logic.

Reply:
What about other potentially valid tokens?  Does it make sense to delete the entire user and all related tokens, forcing them to re-oauth through the entire chain?  The answer may be yes; if so, it’s worth an explicit comment with reasoning.

Reply:
deleteUser only deletes the user information in this connector. All foreign tokens are persisted in the storage controlled by the OAuth connectors this connector is composed with and remain unaffected.

Reply:
Then I’m very confused - wouldn’t deleting all of the foreignOAuthIdentities remove, in the example, both the hyperproof as well as the slack oauth credential?